drivers/ethernet/eth_sam_gmac.c

```diff
@@ -66,41 +67,50 @@ static inline void dcache_is_enabled(void)
 {
 	dcache_enabled = (SCB->CCR & SCB_CCR_DC_Msk);
 }
-static inline void dcache_invalidate(uint32_t addr, uint32_t size)
+static inline void dcache_invalidate(void *addr, uint32_t size)
 {
 	if (!dcache_enabled) {
 		return;
 	}
 
 	/* Make sure it is aligned to 32B */
-	uint32_t start_addr = addr & (uint32_t)~(GMAC_DCACHE_ALIGNMENT - 1);
-	uint32_t size_full = size + addr - start_addr;
+	uint32_t start_addr = (uint32_t)addr & (uint32_t)~(GMAC_DCACHE_ALIGNMENT - 1);
+	uint32_t size_full = size + (uint32_t)addr - start_addr;
 
-	SCB_InvalidateDCache_by_Addr((uint32_t *)start_addr, size_full);
+	SCB_InvalidateDCache_by_Addr((void *)start_addr, size_full);
 }
 
-static inline void dcache_clean(uint32_t addr, uint32_t size)
+static inline void dcache_clean(void *addr, uint32_t size)
 {
 	if (!dcache_enabled) {
 		return;
 	}
 
 	/* Make sure it is aligned to 32B */
-	uint32_t start_addr = addr & (uint32_t)~(GMAC_DCACHE_ALIGNMENT - 1);
-	uint32_t size_full = size + addr - start_addr;
+	uint32_t start_addr = (uint32_t)addr & (uint32_t)~(GMAC_DCACHE_ALIGNMENT - 1);
+	uint32_t size_full = size + (uint32_t)addr - start_addr;
 
-	SCB_CleanDCache_by_Addr((uint32_t *)start_addr, size_full);
+	SCB_CleanDCache_by_Addr((void *)start_addr, size_full);
 }
 #else
 #define dcache_is_enabled()
 #define dcache_invalidate(addr, size)
 #define dcache_clean(addr, size)
 #endif
 
+#ifdef CONFIG_DCACHE
+#undef dcache_invalidate
+#undef dcache_clean
+#define dcache_invalidate sys_cache_data_invd_range
```

Comment:
why not use sys_cache_data_invd_range for all? as the implementation in zephyr for the cortex_m uses SCB_InvalidateDCache_by_Addr internally